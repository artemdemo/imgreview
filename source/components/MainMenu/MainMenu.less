.main-menu {
    padding: 4px 15px;
    background-color: lightgray;
}
