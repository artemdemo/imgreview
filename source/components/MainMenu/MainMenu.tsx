import React from 'react';
import styled from 'styled-components';
import ClearFix from '../Floating/ClearFix';
import * as styleVars from '../../styles/variables';

const MainMenu = styled(ClearFix)`
    padding: 4px 15px;
    background-color: ${styleVars.mainMenuColor};
    z-index: ${styleVars.mainMenuZIndex};
`;

export default MainMenu;
