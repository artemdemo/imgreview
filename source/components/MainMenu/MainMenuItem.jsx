import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import './MainMenuItem.less';

class MainMenuItem extends React.PureComponent {
    render() {
        const { disabled, right, onClick } = this.props;
        const buttonClass = classnames({
            'main-menu-item': true,
            'main-menu-item_disabled': disabled,
            'main-menu-item_right': right,
        });
        return (
            <button
                className={buttonClass}
                disabled={disabled}
                type='button'
                onClick={onClick}
            >
                {this.props.children}
            </button>
        );
    }
}

MainMenuItem.propTypes = {
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
    right: PropTypes.bool,
};

MainMenuItem.defaultProps = {
    onClick: null,
    disabled: null,
    right: false,
};

export default MainMenuItem;
