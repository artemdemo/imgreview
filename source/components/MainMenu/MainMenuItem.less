@import '../../styles/functions';
@import '../../styles/variables';

@horizontal-padding: 4px;
@menu-item-border-color: #b8b8b8;

.main-menu-item {
    .resetButton();
    background-color: @main-menu-color;
    padding: @horizontal-padding 6px;
    border: 1px solid @menu-item-border-color;
    border-radius: 3px;
    float: left;
    margin-right: 5px;
    outline: none;
    display: flex;
    position: relative;
}
    .main-menu-item_disabled {
        opacity: 0.4;
        cursor: not-allowed;
    }

    .main-menu-item__content {
        flex-grow: 1;
        min-width: 30px - 2 * @horizontal-padding;
    }

    .main-menu-item__caret {
        flex-grow: 0;
        padding-left: 5px;
    }

.main-menu-item-submenu {
    position: absolute;
    top: 100%;
    display: none;
    padding-top: 5px;
    z-index: @main-menu-z-index;

    .main-menu-item:hover & {
        display: block;
    }
}
    .main-menu-item-submenu__item {
        padding: 4px 10px;
        border-bottom: 1px dashed @menu-item-border-color;

        &:last-child {
            border-bottom: none;
        }
    }

.main-menu-item-submenu-body {
    background-color: @main-menu-color;
    border: 1px solid @menu-item-border-color;
    border-radius: 3px;
    padding: 3px;
}
