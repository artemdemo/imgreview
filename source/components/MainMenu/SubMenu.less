@import '../../styles/functions';
@import '../../styles/variables';

.submenu {
    background-color: @main-menu-color;
    border: 1px solid @main-menu-item-border-color;
    border-radius: 3px;
    padding: 3px;
}

.submenu-item {
    padding: 4px 10px;
    border-bottom: 1px dashed @main-menu-item-border-color;
    display: flex;
    text-align: left;

    &:last-child {
        border-bottom: none;
    }
}
    .submenu-item__content {
        flex-grow: 1;
    }
    .submenu-item__icon {
        flex-grow: 0;
        padding-left: 10px;
    }
