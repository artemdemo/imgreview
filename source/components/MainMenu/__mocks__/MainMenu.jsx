import React from 'react';

const MainMenu = props => (
    <div
        data-mock='MainMenu'
    >
        {props.children}
    </div>
);

export default MainMenu;
