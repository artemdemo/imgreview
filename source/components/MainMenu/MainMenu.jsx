import React from 'react';
import { connect } from 'react-redux';
import MainMenuItem from './MainMenuItem';
import ColorSelector from '../ColorSelector/ColorSelector';
import OpenImageDialog from '../OpenImageDialog/OpenImageDialog';
import Popup from '../Popup/Popup';
import { saveCanvas } from '../../model/canvas/canvasActions';
import { addArrow, blurShapes, showColorPicker } from '../../model/shapes/shapesActions';

import './MainMenu.less';
import Arrow from '../../canvas/Arrow/Arrow';

const menu = [
    {
        id: 'open-image',
        name: 'Open image',
        icon: 'folder-open-o',
    },
    {
        id: 'save',
        name: 'Save',
        icon: 'floppy-o',
    },
    {
        id: 'vector',
        name: 'Vector',
        icon: 'mouse-pointer',
    },
];

class MainMenu extends React.PureComponent {
    constructor(props) {
        super(props);

        this.openImgDialogRef = React.createRef();
        this.popupRef = React.createRef();
        this.fileNameRef = React.createRef();
    }

    clickOnItem = (item) => {
        const { canvas, shapes, blurShapes, addArrow } = this.props;
        switch (item.id) {
            case 'open-image':
                this.openImgDialogRef.current.openDialog();
                break;
            case 'save':
                blurShapes();
                this.popupRef.current.show();
                break;
            case 'vector':
                const arrow = new Arrow({
                    stroke: shapes.stroke,
                    strokeWidth: shapes.strokeWidth,
                });
                arrow.addToStage(canvas.stage);
                addArrow(arrow);
                break;
            case 'color-selector':
                const { showColorPicker } = this.props;
                showColorPicker();
                break;
        }
    };

    isDisabled = (item) => {
        const { canvas } = this.props;
        switch (item.id) {
            case 'vector':
            case 'save':
                return canvas.image == null;
            default:
                return false;
        }
    };

    renderColorSelector() {
        const { canvas, shapes } = this.props;
        const item = {
            id: 'color-selector',
            name: 'Color selector',
        };
        return (
            <React.Fragment>
                <MainMenuItem
                    item={item}
                    onClick={this.clickOnItem}
                    disabled={canvas.image == null}
                    key={`main-menu-item__${item.id}`}
                >
                    <div
                        className='main-menu-color'
                        style={{
                            backgroundColor: shapes.stroke,
                        }}
                    />
                </MainMenuItem>
                <ColorSelector />
            </React.Fragment>
        );
    }

    render() {
        return (
            <div className='main-menu'>
                {menu.map((item) => {
                    const disabled = this.isDisabled(item);
                    return (
                        <MainMenuItem
                            item={item}
                            onClick={this.clickOnItem}
                            disabled={disabled}
                            key={`main-menu-item__${item.icon}`}
                        />
                    );
                })}
                {this.renderColorSelector()}
                <OpenImageDialog
                    ref={this.openImgDialogRef}
                />
                <Popup
                    ref={this.popupRef}
                    buttons={[
                        {
                            className: 'btn btn-secondary',
                            text: 'Cancel',
                        },
                        {
                            className: 'btn btn-primary',
                            text: 'Save',
                            onClick: () => {
                                const { saveCanvas } = this.props;
                                const { value } = this.fileNameRef.current;
                                if (value !== '') {
                                    saveCanvas(value);
                                }
                            },
                        },
                    ]}
                    showCloseBtn={false}
                >
                    <div className='form-group'>
                        <label htmlFor='saveAs'>Save as</label>
                        <input
                            className='form-control'
                            placeholder='Enter file name'
                            ref={this.fileNameRef}
                            id='saveAs'
                        />
                    </div>
                </Popup>
            </div>
        );
    }
}

export default connect(
    state => ({
        canvas: state.canvas,
        shapes: state.shapes,
    }),
    {
        saveCanvas,
        blurShapes,
        addArrow,
        showColorPicker,
    }
)(MainMenu);
