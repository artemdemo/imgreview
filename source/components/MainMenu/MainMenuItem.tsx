import React from 'react';
import classnames from 'classnames';
import Icon from '../Icon/Icon';
import SubMenu, { TSubmenuData } from './SubMenu';

import './MainMenuItem.less';

type Props = {
    subMenu: TSubmenuData;
    disabled: boolean;
    onClick: (e?: any) => void;
};

class MainMenuItem extends React.PureComponent<Props> {
    static readonly defaultProps = {
        onClick: null,
        disabled: null,
        subMenu: [],
    };

    renderCaret() {
        const { subMenu } = this.props;
        if (subMenu.length > 0) {
            return (
                <span className='main-menu-item__caret'>
                    <Icon name='caret-down' />
                </span>
            );
        }

        return null;
    }

    renderSubMenu() {
        const { subMenu } = this.props;
        if (subMenu.length > 0) {
            return (
                <div className='main-menu-item-submenu'>
                    <SubMenu data={subMenu} />
                </div>
            );
        }

        return null;
    }

    render() {
        const { disabled, onClick } = this.props;
        const buttonClass = classnames({
            'main-menu-item': true,
            'main-menu-item_disabled': disabled,
        });
        return (
            <button
                className={buttonClass}
                disabled={disabled}
                type='button'
                onClick={onClick}
            >
                <span
                    className='main-menu-item__content'
                >
                    {this.props.children}
                </span>
                {this.renderCaret()}
                {this.renderSubMenu()}
            </button>
        );
    }
}

export default MainMenuItem;
