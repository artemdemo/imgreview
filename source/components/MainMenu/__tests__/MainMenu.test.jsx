import React from 'react';
import renderer from 'react-test-renderer';
import MainMenu from '../MainMenu';

describe('MainMenu', () => {
    it('simple render', () => {
        const tree = renderer.create(
            <MainMenu />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
