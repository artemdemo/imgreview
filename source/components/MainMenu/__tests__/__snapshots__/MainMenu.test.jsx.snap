// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainMenu simple render 1`] = `
<div
  className="sc-bdVaJa sc-bwzfXH QZZFO"
/>
`;
